moveit_ros/moveit_servo/src/enforce_limits.cpp

```diff
@@ -69,17 +69,27 @@ double getVelocityScalingFactor(const moveit::core::JointModelGroup* joint_model
 
 }  // namespace
 
-Eigen::ArrayXd enforceVelocityLimits(const moveit::core::JointModelGroup* joint_model_group, double publish_period,
-                                     const Eigen::ArrayXd& delta_theta)
+void enforceVelocityLimits(const moveit::core::JointModelGroup* joint_model_group, const double publish_period,
+                           sensor_msgs::msg::JointState& joint_state)
 {
-  // Convert to joint angle velocities for checking and applying joint specific velocity limits.
-  Eigen::ArrayXd velocity = delta_theta / publish_period;
-
   // Get the velocity scaling factor
+  Eigen::VectorXd velocity =
+      Eigen::Map<Eigen::VectorXd, Eigen::Unaligned>(joint_state.velocity.data(), joint_state.velocity.size());
   double velocity_scaling_factor = getVelocityScalingFactor(joint_model_group, velocity);
 
-  // Scale the resulting detas to avoid violating limits.
-  return velocity_scaling_factor * velocity * publish_period;
+  // Take a smaller step if the velocity scaling factor is less than 1
+  if (velocity_scaling_factor < 1)
+  {
+    Eigen::VectorXd velocity_residuals = (1 - velocity_scaling_factor) * velocity;
+    Eigen::VectorXd positions =
+        Eigen::Map<Eigen::VectorXd, Eigen::Unaligned>(joint_state.position.data(), joint_state.position.size());
+    positions -= velocity_residuals * publish_period;
+
+    velocity *= velocity_scaling_factor;
+    // Back to sensor_msgs type
+    joint_state.velocity = std::vector<double>(velocity.data(), velocity.data() + velocity.size());
+    joint_state.position = std::vector<double>(positions.data(), positions.data() + positions.size());
```

Comment:
It looks like the positions haven't been updated before your fixes. Is that accurate? Modifying the positions with the velocity residuals makes sense to me. I'm only wondering if the resulting position delta is really the product of the computed velocity, considering the smoothing step before velocity enforcement.

Reply:
The positions have been updated prior to this. You can see that in ServoCalcs::internalServoUpdate() --> applyJointUpdate().

The order is:

- apply velocity scaling for collisions or singularities (based on a position command)
- low-pass filter the position command in `applyJointUpdate()`
- calculate velocities in `applyJointUpdate()`
- apply velocity limits
- apply position limits  (I think this is a higher priority than velocity limits, so check it last)

So I think the answer to your question is, it's OK. All calculations and smoothing are done in the position domain, then velocities are calculated and checked at the end.

Reply:
I'll add a comment about this OOP to the code.

Reply:
thanks!